src/Common/src/TypeSystem/Interop/MethodDesc.Interop.cs

```diff
@@ -63,11 +63,51 @@ public struct ParameterMetadata
         public bool HasFieldMarshal { get { return (_attributes & ParameterMetadataAttributes.HasFieldMarshal) == ParameterMetadataAttributes.HasFieldMarshal; } }
 
 
-        public ParameterMetadata(int index, ParameterMetadataAttributes attributes, TypeDesc marshalAs)
+        public ParameterMetadata(int index, ParameterMetadataAttributes attributes, MarshalAsDescriptor marshalAsDescriptor)
         {
             Index = index;
             _attributes = attributes;
-            MarshalAs = marshalAs;       
+            MarshalAsDescriptor = marshalAsDescriptor;
+        }
+    }
+
+    [Flags]
+    public enum NativeType : byte
+    {
+        NATIVE_TYPE_INVALID = 0x0,
+        NATIVE_TYPE_BOOLEAN = 0x2,
+        NATIVE_TYPE_I1 = 0x3,
+        NATIVE_TYPE_U1 = 0x4,
+        NATIVE_TYPE_I2 = 0x5,
+        NATIVE_TYPE_U2 = 0x6,
+        NATIVE_TYPE_I4 = 0x7,
+        NATIVE_TYPE_U4 = 0x8,
+        NATIVE_TYPE_I8 = 0x9,
+        NATIVE_TYPE_U8 = 0xa,
+        NATIVE_TYPE_R4 = 0xb,
+        NATIVE_TYPE_R8 = 0xc,
+        NATIVE_TYPE_LPSTR = 0x14,
+        NATIVE_TYPE_LPWSTR = 0x15,
+        NATIVE_TYPE_INT = 0x1f,
+        NATIVE_TYPE_UINT = 0x20,
+        NATIVE_TYPE_FUNC = 0x26,
+        NATIVE_TYPE_ARRAY = 0x2a,
+        NATIVE_TYPE_LPSTRUCT = 0x2b,    // This is not  defined in Ecma=335(II.23.4)
+        NATIVE_TYPE_MAX = 0x50,         // The value is of this not defined in Ecma-335(II.23.4) either, the one defined in CoreCLR is used here
+    }
+
+    public class MarshalAsDescriptor
+    {
+        NativeType Type;
```

Comment:
Did you mean to make these `public readonly`? Making these properties would be nice too.